### PR TITLE
feat: stonedb initial support

### PR DESCRIPTION
**Pull request by ZhengLin-Li**
initial support of StoneDB

there are still testing errors, but since this is a WIP PR, so let's ignore these errors first?
```
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[135,36] The method getColumnType(java.lang.String) from the type sqlancer.stonedb.StoneDBSchema is never used locally
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[179,27] The value of the field sqlancer.stonedb.StoneDBSchema.StoneDBColumn.precision is not used
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant BINARY needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant BLOB needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant CHAR needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant DATE needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant DATETIME needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant DECIMAL needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant LONGBLOB needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant LONGTEXT needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant MEDIUMBLOB needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant MEDIUMTEXT needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant TEXT needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant TIME needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant TIMESTAMP needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant TINYBLOB needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant TINYTEXT needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant VARBINARY needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant VARCHAR needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType
D:\Documents\Projects\sqlancer\src\sqlancer\stonedb\StoneDBSchema.java:[215,21] The enum constant YEAR needs a corresponding case label in this enum switch on sqlancer.stonedb.StoneDBSchema.StoneDBDataType

```

### CODE REVIEW

File: src/sqlancer/stonedb/StoneDBOptions.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import sqlancer.stonedb.StoneDBProvider.StoneDBGlobalState;
+import sqlancer.stonedb.oracle.StoneDBNoRECOracle;
+import sqlancer.stonedb.oracle.StoneDBTLPOracle;
+import sqlancer.stonedb.StoneDBOptions.StoneDBOracleFactory;
+
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.List;
+
+@Parameters(separators = "=", commandDescription = "StoneDB (default host: " + MySQLOptions.DEFAULT_HOST
```
Comment (bajinsheng) [inline src/sqlancer/stonedb/StoneDBOptions.java:18]: Here, I think you should create a StoneDBOption, as it defines DBMS-specific parameters.

Comment (bajinsheng) [reply src/sqlancer/stonedb/StoneDBOptions.java:18]: Oh, I just noticed that you have a StoneDBOption class, but do not import anything from MySQLOption class. Define it in StoneDBOption class.

File: src/sqlancer/stonedb/StoneDBSchema.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        // todo: support more data type
+        default:
+            throw new AssertionError(typeString);
+        }
+    }
+
+    private static List<String> getTableNames(SQLConnection con) throws SQLException {
+        List<String> tableNames = new ArrayList<>();
+        try (Statement s = con.createStatement()) {
+            try (ResultSet rs = s.executeQuery("SELECT * FROM sqlite_master WHERE type='table' or type='view'")) {
```
Comment (bajinsheng) [inline src/sqlancer/stonedb/StoneDBSchema.java:167]: Is the SQL corrected? Isn't it `SHOW TABLE`?

File: src/sqlancer/stonedb/StoneDBSchema.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        public StoneDBCompositeDataType(StoneDBDataType dataType, int size) {
+            this.dataType = dataType;
+            this.size = size;
+        }
+
+        public StoneDBCompositeDataType(StoneDBDataType dataType) {
+            this.dataType = dataType;
+            size = -1;
+            switch (dataType) {
+            case TINYINT:
```
Comment (bajinsheng) [inline src/sqlancer/stonedb/StoneDBSchema.java:217]: These data types are implemented differently in MySQL, but seem fine-grained and correct with reference to the doc: https://stonedb.io/docs/SQL-reference/data-types?_highlight=data .

Have you checked with the StoneDB doc? If so, it is good.

File: src/sqlancer/stonedb/StoneDBSchema.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class StoneDBSchema extends AbstractSchema<StoneDBProvider.StoneDBGlobalState, StoneDBSchema.StoneDBTable> {
+
+    public enum StoneDBDataType {
+
+        TINYINT, SMALLINT, MEDIUMINT, INT, BIGINT, FLOAT, DOUBLE, DECIMAL, YEAR, TIME, DATE, DATETIME, TIMESTAMP, CHAR,
```
Comment (bajinsheng) [inline src/sqlancer/stonedb/StoneDBSchema.java:22]: For these errors of undefined enum, I would suggest you can simply add definitions of them in corresponding functions to make the PR self-contained.

**Review by bajinsheng — CHANGES_REQUESTED** [review]

**Review by bajinsheng — APPROVED** [review]

**Comment by mrigger** [comment]
It's strange, it seems they are stuck. Not sure if pushing again might help?

**Comment by oraluben** [comment]
Looks like a github issue, I think close and open a new PR would work.

**Comment by ZhengLin-Li** [comment]
will close and reopen

**Pull request closed without merging**